File: scripts/inspect-mindmap-dom.py
#!/usr/bin/env python3
"""
Inspect Mermaid mindmap DOM structure using Selenium
"""

import sys
import time
from pathlib import Path

try:
    from selenium import webdriver
    from selenium.webdriver.common.by import By
    from selenium.webdriver.support.ui import WebDriverWait
    from selenium.webdriver.support import expected_conditions as EC
    from selenium.webdriver.chrome.options import Options
except ImportError:
    print("ERROR: selenium not installed")
    print("Install with: pip3 install selenium")
    sys.exit(1)

def inspect_mindmap_dom(html_path):
    """Open HTML file and inspect the mindmap DOM structure"""
    
    # Setup Chrome in headless mode
    chrome_options = Options()
    chrome_options.add_argument('--headless')
    chrome_options.add_argument('--disable-gpu')
    
    try:
        driver = webdriver.Chrome(options=chrome_options)
    except Exception as e:
        print(f"ERROR: Could not start Chrome driver: {e}")
        print("You may need to install chromedriver")
        sys.exit(1)
    
    try:
        # Load the page
        file_url = f"file://{html_path}"
        driver.get(file_url)
        
        # Wait for the diagram to render
        wait = WebDriverWait(driver, 10)
        wait.until(EC.presence_of_element_located((By.CSS_SELECTOR, ".mindmap-node")))
        
        # Give extra time for complete rendering
        time.sleep(1)
        
        # Get all mindmap nodes
        nodes = driver.find_elements(By.CSS_SELECTOR, ".mindmap-node")
        
        print("=== MINDMAP DOM INSPECTION ===\n")
        print(f"Total nodes found: {len(nodes)}\n")
        
        for idx, node in enumerate(nodes):
            text_el = node.find_element(By.CSS_SELECTOR, "text")
            text_content = text_el.text.strip()
            
            # Focus on non-root nodes, especially "Preview"
            if text_content == "Preview" or (idx >= 3 and idx <= 6):
                print(f"\n=== NODE: {text_content} (index {idx}) ===")
                print(f"Parent g classes: {node.get_attribute('class')}")
                print(f"\n<text> element:")
                print(f"  id: {text_el.get_attribute('id')}")
                print(f"  class: {text_el.get_attribute('class')}")
                print(f"  transform: {text_el.get_attribute('transform')}")
                print(f"  text-anchor: {text_el.get_attribute('text-anchor')}")
                print(f"  dominant-baseline: {text_el.get_attribute('dominant-baseline')}")
                print(f"  alignment-baseline: {text_el.get_attribute('alignment-baseline')}")
                print(f"  y: {text_el.get_attribute('y')}")
                print(f"  dy: {text_el.get_attribute('dy')}")
                print(f"  style: {text_el.get_attribute('style')}")
                
                # Get tspan elements
                tspans = text_el.find_elements(By.CSS_SELECTOR, "tspan")
                print(f"\n  <tspan> elements ({len(tspans)} total):")
                for t_idx, tspan in enumerate(tspans):
                    print(f"    tspan[{t_idx}]:")
                    print(f"      class: {tspan.get_attribute('class')}")
                    print(f"      x: {tspan.get_attribute('x')}")
                    print(f"      y: {tspan.get_attribute('y')}")
                    print(f"      dy: {tspan.get_attribute('dy')}")
                    print(f"      dominant-baseline: {tspan.get_attribute('dominant-baseline')}")
                    print(f"      alignment-baseline: {tspan.get_attribute('alignment-baseline')}")
                    print(f"      text-anchor: {tspan.get_attribute('text-anchor')}")
                    print(f"      style: {tspan.get_attribute('style')}")
                    print(f'      textContent: "{tspan.text}"')
                
                print(f"\n  Full outerHTML:")
                print(text_el.get_attribute('outerHTML'))
        
    finally:
        driver.quit()

if __name__ == "__main__":
    html_path = sys.argv[1] if len(sys.argv) > 1 else "/Users/soleilyu/Code/self/md2jpeg/scripts/.tmp/dom-inspector.html"
    html_path = Path(html_path).absolute()
    
    if not html_path.exists():
        print(f"ERROR: File not found: {html_path}")
        sys.exit(1)
    
    inspect_mindmap_dom(html_path)

File: scripts/inspect-mindmap-dom.js
#!/usr/bin/env node

/**
 * Script to inspect Mermaid mindmap DOM structure
 * This script opens the HTML file in a headless browser and extracts
 * the exact DOM structure of mindmap node labels.
 */

const fs = require('fs');
const path = require('path');

// Read the HTML file
const htmlPath = process.argv[2] || path.join(__dirname, '.tmp', 'latest.html');
const htmlContent = fs.readFileSync(htmlPath, 'utf-8');

// Extract the mermaid source
const mermaidMatch = htmlContent.match(/<pre class="mermaid">([\s\S]*?)<\/pre>/);
if (!mermaidMatch) {
  console.error('Could not find mermaid diagram in HTML');
  process.exit(1);
}

const mermaidSource = mermaidMatch[1].trim();

// Create a simple HTML page that will render and inspect
const inspectorHTML = `
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <script src="https://cdn.jsdelivr.net/npm/mermaid@10.9.5/dist/mermaid.min.js"></script>
</head>
<body>
  <div id="diagram"></div>
  <script>
    mermaid.initialize({
      theme: "base",
      startOnLoad: false,
      securityLevel: "strict",
      themeVariables: {
        fontFamily: "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", sans-serif",
        fontSize: "14px",
        darkMode: true,
        primaryColor: "#1e293b",
        primaryTextColor: "#e2e8f0",
        primaryBorderColor: "#475569",
        lineColor: "#60a5fa",
        secondaryColor: "#0f172a",
        tertiaryColor: "#1e293b",
        background: "#0f172a",
        mainBkg: "#1e293b",
        textColor: "#e2e8f0",
        noteBkgColor: "#1e293b",
        noteTextColor: "#94a3b8",
        noteBorderColor: "#475569",
        cScale0: "#243352",
        cScaleLabel0: "#ffffff"
      }
    });
    
    const source = \`${mermaidSource}\`;
    
    mermaid.render("diagram-svg", source).then(result => {
      document.getElementById('diagram').innerHTML = result.svg;
      
      // Wait for rendering to complete
      setTimeout(() => {
        const svg = document.querySelector('#diagram svg');
        if (!svg) {
          console.log('ERROR: SVG not found');
          return;
        }
        
        const nodes = svg.querySelectorAll('.mindmap-node');
        console.log('=== MINDMAP DOM INSPECTION ===\\n');
        console.log('Total nodes found:', nodes.length, '\\n');
        
        nodes.forEach((node, idx) => {
          const textEl = node.querySelector('text');
          if (!textEl) return;
          
          const textContent = textEl.textContent.trim();
          
          // Focus on non-root nodes, especially "Preview"
          if (textContent === 'Preview' || (idx >= 3 && idx <= 6)) {
            console.log('\\n=== NODE:', textContent, '(index', idx + ') ===');
            console.log('Parent g classes:', node.getAttribute('class'));
            console.log('\\n<text> element:');
            console.log('  id:', textEl.getAttribute('id'));
            console.log('  class:', textEl.getAttribute('class'));
            console.log('  transform:', textEl.getAttribute('transform'));
            console.log('  text-anchor:', textEl.getAttribute('text-anchor'));
            console.log('  dominant-baseline:', textEl.getAttribute('dominant-baseline'));
            console.log('  alignment-baseline:', textEl.getAttribute('alignment-baseline'));
            console.log('  y:', textEl.getAttribute('y'));
            console.log('  dy:', textEl.getAttribute('dy'));
            console.log('  style:', textEl.getAttribute('style'));
            
            const tspans = textEl.querySelectorAll('tspan');
            console.log('\\n  <tspan> elements (' + tspans.length + ' total):');
            tspans.forEach((tspan, tIdx) => {
              console.log('    tspan[' + tIdx + ']:');
              console.log('      class:', tspan.getAttribute('class'));
              console.log('      x:', tspan.getAttribute('x'));
              console.log('      y:', tspan.getAttribute('y'));
              console.log('      dy:', tspan.getAttribute('dy'));
              console.log('      dominant-baseline:', tspan.getAttribute('dominant-baseline'));
              console.log('      alignment-baseline:', tspan.getAttribute('alignment-baseline'));
              console.log('      text-anchor:', tspan.getAttribute('text-anchor'));
              console.log('      style:', tspan.getAttribute('style'));
              console.log('      textContent:', '"' + tspan.textContent + '"');
            });
            
            console.log('\\n  Full outerHTML:');
            console.log(textEl.outerHTML);
          }
        });
        
      }, 100);
    }).catch(err => {
      console.error('ERROR:', err.message);
    });
  </script>
</body>
</html>
`;

// Try using jsdom if available
try {
  const { JSDOM } = require('jsdom');
  
  const dom = new JSDOM(inspectorHTML, {
    runScripts: "dangerously",
    resources: "usable",
    beforeParse(window) {
      // Set up console to capture output
      window.console = console;
    }
  });
  
  // Keep process alive for async operations
  setTimeout(() => {
    process.exit(0);
  }, 5000);
  
} catch (err) {
  // jsdom not available, output the HTML for manual inspection
  console.log('jsdom not available. Please install it with: npm install jsdom');
  console.log('\\nAlternatively, save this HTML and open in a browser:');
  console.log('\\n' + inspectorHTML);
  
  const outputPath = path.join(__dirname, '.tmp', 'dom-inspector.html');
  fs.writeFileSync(outputPath, inspectorHTML);
  console.log('\\nHTML saved to:', outputPath);
  console.log('Open it in a browser and check the console for output.');
}
